File: client-sdk/typescript/src/service.ts
// Server-side protocol-compliant agent per the §12 implementation checklist.
//
// `AgentService` handles the boilerplate every protocol-compliant agent
// needs:
//   - Registers as the NATS micro service `agents` (§3.1) with metadata
//     `{agent, owner, protocol_version: "0.3", session?}` (§3.2).
//   - Adds a `prompt` endpoint at `agents.prompt.<agent>.<owner>.<name>`
//     with queue group `"agents"` (§3.3) and metadata `{max_payload,
//     attachments_ok}` (§2.1).
//   - Adds a `status` endpoint at `agents.status.<agent>.<owner>.<name>`
//     (v0.3 §-TBD) that replies with a freshly-built §8.3 heartbeat
//     payload. Same queue group as `prompt`.
//   - Publishes heartbeats on `agents.hb.<agent>.<owner>.<name>` every
//     `heartbeatIntervalS` seconds (§8.1, §8.2). First heartbeat is
//     emitted immediately after registration so callers discovering via
//     `$SRV.INFO.agents` can observe liveness without waiting a full
//     interval (§8.5).
//   - Runs a per-request keep-alive task that emits
//     `{"type":"status","data":"ack"}` every `keepaliveIntervalS`
//     seconds while the prompt handler is running, so callers using a
//     stream inactivity timeout (the SDK default is 60 s) don't fire on
//     slow handlers. Cancelled before the §6.5 stream terminator so the
//     ack stream never extends past the terminator.
//   - Emits the spec-mandated empty-body no-headers terminator after
//     every prompt — successful or errored (§6.5, §9.3).
//   - Translates handler exceptions into `Nats-Service-Error-Code: 500`
//     responses; envelope decode failures into `400` (§9.1).
//
// Mirrors the Python SDK's `AgentService` (`client-sdk/python/src/synadia_ai/agents/service.py`)
// — wire-equivalent behaviour, idiomatic TS API.

import type { NatsConnection } from "@nats-io/nats-core";
import { Svcm, type Service, type ServiceMsg } from "@nats-io/services";

import { ProtocolError } from "./errors.js";
import { newInbox } from "./internal/inbox.js";
import {
  PROMPT_QUEUE_GROUP,
  SERVICE_NAME,
  STATUS_ENDPOINT_NAME,
  STATUS_QUEUE_GROUP,
} from "./internal/service-name.js";
import { PROMPT_ENDPOINT_NAME } from "./discovery/endpoint-info.js";
import {
  buildHeartbeatPayload,
  encodeHeartbeatPayload,
} from "./heartbeat/payload.js";
import {
  decodeEnvelope,
  encodeBase64,
  encodeEnvelope,
  type RequestAttachment,
  type RequestEnvelope,
} from "./prompt/envelope.js";
import {
  encodeChunk,
  type Chunk,
  type QueryChunk,
  type StatusChunk,
} from "./stream/chunk-encoder.js";
import { AgentSubject } from "./subjects.js";
import { SDK_PROTOCOL_VERSION } from "./version.js";

/** §3.2 + §11.1: `metadata.protocol_version` is MAJOR.MINOR only. */
const PROTOCOL_VERSION_STRING = `${SDK_PROTOCOL_VERSION.major}.${SDK_PROTOCOL_VERSION.minor}`;

/** Default §2.1 prompt-endpoint metadata. */
export const DEFAULT_MAX_PAYLOAD = "1MB";
export const DEFAULT_ATTACHMENTS_OK = true;

/** Default §8.2 heartbeat cadence — matches Python (and pre-v0.3 TS reference agent). */
export const DEFAULT_HEARTBEAT_INTERVAL_S = 30;

/**
 * Default per-request keep-alive cadence. Every TS reference harness
 * (`agents/pi/`, `agents/claude-code/`, `agents/openclaw/`) emits
 * `{"type":"status","data":"ack"}` periodically while a request is in
 * flight so callers using a stream inactivity timeout (§6.6 — TS SDK
 * default 60 s) don't fire on slow handlers. Defaults to 30 s; pass
 * `null` to disable.
 */
export const DEFAULT_KEEPALIVE_INTERVAL_S = 30;

/** Default `service.version` advertised in `$SRV.INFO`. */
const DEFAULT_VERSION = "0.0.1";

export interface AgentServiceOptions {
  /** A pre-connected `NatsConnection`. Caller retains ownership. */
  readonly nc: NatsConnection;
  /** §3.2 `metadata.agent` — canonical harness identifier (e.g. `"claude-code"`). */
  readonly agent: string;
  /** §3.2 `metadata.owner` — operator / account namespace. */
  readonly owner: string;
  /** Instance name — 5th subject token (§2 v0.3). */
  readonly name: string;
  /**
   * Optional §5.6 envelope-level conversation label. Set when the agent
   * advertises a session label in `$SRV.INFO.metadata.session` and emits
   * it on every heartbeat — used by harnesses (e.g. Hermes) that
   * multiplex multiple conversations over a single subject.
   */
  readonly session?: string;
  /** Human-readable service description. Defaults to `"<agent> agent <name>"`. */
  readonly description?: string;
  /** Harness semver (`service.version`). Defaults to `"0.0.1"`. */
  readonly version?: string;
  /** §2.1 `max_payload`. Defaults to `"1MB"`. */
  readonly maxPayload?: string;
  /** §2.1 `attachments_ok`. Defaults to `true`. */
  readonly attachmentsOk?: boolean;
  /** §8.2 heartbeat cadence in seconds. Must be > 0. Defaults to 30. */
  readonly heartbeatIntervalS?: number;
  /**
   * Per-request keep-alive cadence in seconds. Must be > 0, or `null` to
   * disable (e.g. when the handler emits its own status chunks at a finer
   * cadence). Defaults to 30.
   */
  readonly keepaliveIntervalS?: number | null;
  /** Extra metadata keys merged into the service metadata (forward-compat). */
  readonly extraMetadata?: Readonly<Record<string, string>>;
}

export type PromptHandler = (
  envelope: RequestEnvelope,
  response: PromptResponse,
) => Promise<void> | void;

/** A "non-empty subset of `crypto.randomUUID()`-style tokens" — used for query ids. */
function randomId(): string {
  return crypto.randomUUID().replace(/-/g, "");
}

/**
 * Server-side handle given to a {@link PromptHandler} for emitting response
 * chunks back to the caller. The {@link AgentService} owns stream
 * termination — handlers `send(...)` zero or more chunks and return.
 *
 * Distinct from the client-side `PromptStream` (`@synadia-ai/agents`'s
 * `Agent.prompt(...)` return value). Same conceptual stream, opposite ends.
 */
export class PromptResponse {
  readonly #msg: ServiceMsg;
  readonly #nc: NatsConnection;

  constructor(msg: ServiceMsg, nc: NatsConnection) {
    this.#msg = msg;
    this.#nc = nc;
  }

  /**
   * Publish one chunk to the caller's reply subject.
   *
   * A `string` is wrapped in a `ResponseChunk` and emitted as the §6.3
   * bare-string form. §6.2 forbids plain-text shorthand on the response
   * side — every non-terminating chunk is a JSON object with a `type`
   * field. Returns a Promise for forward-compat with backpressure/flush
   * variants; today the underlying `respond` is fire-and-forget.
   */
  send(chunk: string | Chunk): Promise<void> {
    const c: Chunk = typeof chunk === "string" ? { type: "response", text: chunk } : chunk;
    this.#msg.respond(encodeChunk(c));
    return Promise.resolve();
  }

  /**
   * Ask the caller a mid-stream question and await a single reply (§7).
   *
   * Allocates a fresh reply inbox, publishes a `query` chunk into the
   * response stream, and waits up to `timeoutMs` milliseconds. The
   * response stream stays open across the round-trip — the caller keeps
   * iterating the prompt while the handler awaits here.
   *
   * Throws on timeout — handlers decide whether to abort the stream or
   * proceed with a default per §7.3.
   */
  async ask(
    prompt: string | RequestEnvelope,
    opts: { readonly timeoutMs: number; readonly attachments?: ReadonlyArray<RequestAttachment> },
  ): Promise<RequestEnvelope> {
    const promptText = typeof prompt === "string" ? prompt : prompt.prompt;
    const baseAttachments = typeof prompt === "string" ? undefined : prompt.attachments;
    const merged: RequestAttachment[] = [
      ...(baseAttachments ?? []),
      ...(opts.attachments ?? []),
    ];

    const replySubject = newInbox();
    const sub = this.#nc.subscribe(replySubject, { max: 1 });
    await this.#nc.flush();

    const queryChunk: QueryChunk = {
      type: "query",
      id: randomId(),
      replySubject,
      prompt: promptText,
      ...(merged.length > 0
        ? {
            attachments: merged.map((a) => ({
              filename: a.filename,
              content: encodeBase64(a.content),
            })),
          }
        : {}),
    };
    this.#msg.respond(encodeChunk(queryChunk));

    let timer: ReturnType<typeof setTimeout> | undefined;
    try {
      const next = (async (): Promise<RequestEnvelope> => {
        for await (const m of sub) {
          // First message wins (max:1 subscription).
          return decodeEnvelope(m.data);
        }
        throw new Error(
          `query ${queryChunk.id} reply subscription closed before any reply arrived`,
        );
      })();
      const timed = new Promise<never>((_, reject) => {
        timer = setTimeout(
          () =>
            reject(
              new Error(
                `query ${queryChunk.id} on ${replySubject} timed out after ${opts.timeoutMs}ms`,
              ),
            ),
          opts.timeoutMs,
        );
      });
      return await Promise.race([next, timed]);
    } finally {
      if (timer !== undefined) clearTimeout(timer);
      sub.unsubscribe();
    }
  }
}

/** A protocol-compliant agent (§12). */
export class AgentService {
  readonly #options: AgentServiceOptions;
  readonly #subject: AgentSubject;
  readonly #heartbeatIntervalS: number;
  readonly #keepaliveIntervalS: number | null;
  #handler: PromptHandler | null = null;
  #service: Service | null = null;
  #heartbeatTimer: ReturnType<typeof setInterval> | null = null;

  constructor(options: AgentServiceOptions) {
    const heartbeatIntervalS = options.heartbeatIntervalS ?? DEFAULT_HEARTBEAT_INTERVAL_S;
    if (heartbeatIntervalS <= 0) {
      throw new Error("AgentService: heartbeatIntervalS must be > 0 (heartbeat is mandatory in v0.3)");
    }
    const keepaliveIntervalS =
      options.keepaliveIntervalS === undefined ? DEFAULT_KEEPALIVE_INTERVAL_S : options.keepaliveIntervalS;
    if (keepaliveIntervalS !== null && keepaliveIntervalS <= 0) {
      throw new Error(
        "AgentService: keepaliveIntervalS must be > 0 or null (null disables keep-alive)",
      );
    }

    this.#options = options;
    this.#subject = AgentSubject.new(options.agent, options.owner, options.name);
    this.#heartbeatIntervalS = heartbeatIntervalS;
    this.#keepaliveIntervalS = keepaliveIntervalS;
  }

  /** The validated subject for this agent. */
  get subject(): AgentSubject {
    return this.#subject;
  }

  /** The service id assigned by `@nats-io/services` once started. Matches heartbeat `instance_id`. */
  get instanceId(): string {
    if (!this.#service) {
      throw new Error("AgentService.instanceId: service not started");
    }
    return this.#service.info().id;
  }

  /** Register the prompt handler. Must be called before {@link start}. */
  onPrompt(handler: PromptHandler): void {
    this.#handler = handler;
  }

  async start(): Promise<void> {
    if (this.#handler === null) {
      throw new Error("AgentService.start: register a prompt handler via onPrompt() first");
    }
    if (this.#service !== null) {
      throw new Error("AgentService.start: already started");
    }

    const svcm = new Svcm(this.#options.nc);
    const metadata: Record<string, string> = {
      agent: this.#subject.agent,
      owner: this.#subject.owner,
      protocol_version: PROTOCOL_VERSION_STRING,
      ...this.#options.extraMetadata,
    };
    if (this.#options.session !== undefined) {
      metadata["session"] = this.#options.session;
    }

    this.#service = await svcm.add({
      name: SERVICE_NAME,
      version: this.#options.version ?? DEFAULT_VERSION,
      description:
        this.#options.description ?? `${this.#subject.agent} agent ${this.#subject.name}`,
      metadata,
    });

    this.#service.addEndpoint(PROMPT_ENDPOINT_NAME, {
      subject: this.#subject.prompt,
      queue: PROMPT_QUEUE_GROUP,
      handler: (err, msg) => {
        if (err) return;
        void this.#dispatchPrompt(msg);
      },
      metadata: {
        max_payload: this.#options.maxPayload ?? DEFAULT_MAX_PAYLOAD,
        attachments_ok: (this.#options.attachmentsOk ?? DEFAULT_ATTACHMENTS_OK) ? "true" : "false",
      },
    });

    this.#service.addEndpoint(STATUS_ENDPOINT_NAME, {
      subject: this.#subject.status,
      queue: STATUS_QUEUE_GROUP,
      handler: (err, msg) => {
        if (err) return;
        this.#dispatchStatus(msg);
      },
    });

    this.#startHeartbeats();
  }

  async stop(): Promise<void> {
    if (this.#heartbeatTimer !== null) {
      clearInterval(this.#heartbeatTimer);
      this.#heartbeatTimer = null;
    }
    if (this.#service !== null) {
      await this.#service.stop();
      this.#service = null;
    }
  }

  #startHeartbeats(): void {
    const publish = (): void => {
      const service = this.#service;
      if (!service) return;
      const payload = buildHeartbeatPayload(
        this.#subject,
        this.#heartbeatIntervalS,
        service.info().id,
        this.#options.session !== undefined ? { session: this.#options.session } : {},
      );
      this.#options.nc.publish(this.#subject.heartbeat, encodeHeartbeatPayload(payload));
    };
    publish();
    this.#heartbeatTimer = setInterval(publish, this.#heartbeatIntervalS * 1000);
    // Allow the Node process to exit even if the timer is still active.
    this.#heartbeatTimer.unref?.();
  }

  #dispatchStatus(msg: ServiceMsg): void {
    const service = this.#service;
    if (!service) return;
    try {
      const payload = buildHeartbeatPayload(
        this.#subject,
        this.#heartbeatIntervalS,
        service.info().id,
        this.#options.session !== undefined ? { session: this.#options.session } : {},
      );
      msg.respond(encodeHeartbeatPayload(payload));
    } catch (err) {
      try {
        msg.respondError(500, sanitizeErrorDesc(`status handler error: ${(err as Error).message}`));
      } catch {
        /* connection may already be gone */
      }
    }
  }

  async #dispatchPrompt(msg: ServiceMsg): Promise<void> {
    const handler = this.#handler;
    if (!handler) return; // start() rejects this path; defensive

    let envelope: RequestEnvelope;
    try {
      envelope = decodeEnvelope(msg.data);
    } catch (err) {
      const desc = err instanceof ProtocolError ? err.message : "malformed prompt envelope";
      try {
        msg.respondError(400, sanitizeErrorDesc(desc));
      } catch {
        /* connection may already be gone */
      }
      // §9.3 — the error frame is NOT the terminator; emit one explicitly.
      tryRespondTerminator(msg);
      return;
    }

    const response = new PromptResponse(msg, this.#options.nc);

    // Per-request keep-alive: while the handler runs, emit ack chunks so
    // callers using a stream inactivity timeout (§6.6) don't fire.
    let keepaliveTimer: ReturnType<typeof setInterval> | null = null;
    if (this.#keepaliveIntervalS !== null) {
      const intervalMs = this.#keepaliveIntervalS * 1000;
      const ack: StatusChunk = { type: "status", status: "ack" };
      const ackBytes = encodeChunk(ack);
      keepaliveTimer = setInterval(() => {
        try {
          msg.respond(ackBytes);
        } catch {
          // best-effort; clear so we don't keep firing on a dead request
          if (keepaliveTimer) clearInterval(keepaliveTimer);
          keepaliveTimer = null;
        }
      }, intervalMs);
      keepaliveTimer.unref?.();
    }

    const stopKeepalive = (): void => {
      if (keepaliveTimer !== null) {
        clearInterval(keepaliveTimer);
        keepaliveTimer = null;
      }
    };

    try {
      await handler(envelope, response);
    } catch (err) {
      // Stop keep-alive BEFORE the §9 error frame so an ack chunk can't
      // race in between the error and the terminator.
      stopKeepalive();
      try {
        msg.respondError(500, sanitizeErrorDesc(`handler error: ${(err as Error).message}`));
      } catch {
        /* connection may already be gone */
      }
    } finally {
      stopKeepalive();
      // §6.5 + §9.3: every stream — successful or errored — ends with a
      // zero-byte body message that carries NO NATS headers.
      tryRespondTerminator(msg);
    }
  }
}

function tryRespondTerminator(msg: ServiceMsg): void {
  try {
    msg.respond(new Uint8Array(0));
  } catch {
    /* connection may already be gone */
  }
}

// NATS message headers are single-line, so any description passed to
// `respondError` MUST be stripped of newlines or the server will truncate
// subsequent headers. 200 chars is plenty for §9.1; richer context belongs
// in the JSON body per §9.1.
const MAX_ERROR_DESC_LEN = 200;

function sanitizeErrorDesc(desc: string): string {
  const flat = desc
    .split(/\r?\n/)
    .map((s) => s.trim())
    .filter((s) => s.length > 0)
    .join(" | ");
  if (flat.length > MAX_ERROR_DESC_LEN) {
    return flat.slice(0, MAX_ERROR_DESC_LEN - 3) + "...";
  }
  return flat;
}

// Re-export for `service.ts` consumers — agent authors typically import
// envelope/chunk types from here rather than from the deeper paths.
export type { RequestEnvelope, RequestAttachment } from "./prompt/envelope.js";
export type { Chunk, ResponseChunk, StatusChunk, QueryChunk } from "./stream/chunk-encoder.js";
export { encodeEnvelope };

File: client-sdk/python/src/synadia_ai/agents/service.py
"""Server-side protocol-compliant agent per the §12 implementation checklist.

An :class:`AgentService` registers as a NATS micro service with the
protocol's required metadata, serves the prompt inbox, and publishes
heartbeats at the configured interval. Agent authors register a prompt
handler via :meth:`AgentService.on_prompt`; the handler receives the
decoded envelope and a :class:`PromptStream` on which to emit response
chunks.

This file hosts the **server side** (what Hermes / claude-code / pi
embed). The **client side** lives in :mod:`synadia_ai.agents.agent` as a
:class:`~synadia_ai.agents.agent.Agent` returned from
:meth:`~synadia_ai.agents.agents.Agents.discover`.
"""

from __future__ import annotations

import asyncio
import contextlib
import uuid
from collections.abc import Awaitable, Callable
from importlib.metadata import PackageNotFoundError
from importlib.metadata import version as _pkg_version
from typing import TYPE_CHECKING

from nats.micro import ServiceConfig, add_service
from nats.micro.service import EndpointConfig

from ._bytes import parse_human_bytes
from ._logging import get_logger
from .discovery import (
    PROMPT_ENDPOINT_NAME,
    PROMPT_QUEUE_GROUP,
    SERVICE_NAME,
    STATUS_ENDPOINT_NAME,
    STATUS_QUEUE_GROUP,
)
from .envelope import Attachment, Envelope, decode
from .errors import ProtocolError, QueryTimeout
from .heartbeat import build_heartbeat_payload, run_publisher
from .messages import (
    Chunk,
    QueryChunk,
    ResponseChunk,
    StatusChunk,
    encode_chunk,
)
from .subjects import AgentSubject

if TYPE_CHECKING:
    from nats.aio.client import Client as NATSClient
    from nats.micro.service import Request, Service


log = get_logger(__name__)

PromptHandler = Callable[["Envelope", "PromptStream"], Awaitable[None]]

# §3.2 + §11.1: metadata.protocol_version is MAJOR.MINOR only.
_PROTOCOL_VERSION = "0.3"


def _resolve_sdk_version() -> str:
    """Read the installed package version so pyproject.toml is the single source.

    Falls back to ``"0.0.0+unknown"`` when the package is not installed
    (e.g. running from a source tree without ``uv sync``) — `$SRV.INFO`
    still emits a syntactically valid version field.
    """
    try:
        return _pkg_version("synadia-ai-agents")
    except PackageNotFoundError:
        return "0.0.0+unknown"


_SDK_VERSION = _resolve_sdk_version()

# §2.1: prompt endpoint metadata defaults. Spec's own example uses 1MB;
# attachments_ok defaults to True so envelopes with `attachments` just work
# out of the box. Agents MAY override via `AgentService(max_payload=..., ...)`.
DEFAULT_MAX_PAYLOAD = "1MB"
DEFAULT_ATTACHMENTS_OK = True

# §6.4: callers may wire a stream-inactivity timeout — the TS SDK defaults to
# 60 s — so an agent that goes quiet for too long looks dead even if its
# handler is still working. To avoid that, every TS reference harness
# (`agents/pi/`, `agents/claude-code/`, `agents/openclaw/`) emits
# `{type:"status",data:"ack"}` periodically while a request is in flight. We
# match that behaviour by default; agent authors who don't want it pass
# `keepalive_interval_s=None`.
DEFAULT_KEEPALIVE_INTERVAL_S: float = 30.0


class PromptStream:
    """Handle given to a prompt handler for emitting response chunks.

    The :class:`AgentService` owns stream termination (empty-payload
    terminator per §6.5). Handlers should ``send(...)`` zero or more
    chunks and return; raising an exception converts to a service error
    per §9.
    """

    def __init__(
        self,
        request: Request,
        nc: NATSClient,
    ) -> None:
        self._request = request
        self._nc = nc

    async def send(self, chunk: str | Chunk) -> None:
        """Publish one chunk to the caller's reply subject.

        A ``str`` is wrapped in a :class:`ResponseChunk` and emitted as the
        §6.3 bare-string form ``{"type":"response","data":"<text>"}``. §6.2
        explicitly forbids plain-text shorthand on the response side — every
        non-terminating chunk MUST be a JSON object with a ``type`` field.
        """
        if isinstance(chunk, str):
            payload = encode_chunk(ResponseChunk(text=chunk))
        elif isinstance(chunk, ResponseChunk | StatusChunk | QueryChunk):
            payload = encode_chunk(chunk)
        else:
            raise TypeError(f"unsupported chunk type: {type(chunk).__name__}")
        await self._request.respond(payload)

    async def ask(
        self,
        prompt: str | Envelope,
        *,
        timeout: float,
        attachments: list[Attachment] | None = None,
    ) -> Envelope:
        """Ask the caller a mid-stream question and await the reply (§7).

        Allocates a fresh reply inbox, publishes a ``query`` chunk into the
        response stream, and waits for exactly one reply. The response stream
        stays open across the round-trip — the caller keeps iterating the
        prompt's async iterator while the handler awaits here.

        ``prompt`` is either a bare string or an :class:`Envelope` whose
        ``prompt`` + ``attachments`` become the query's fields. Raises
        :class:`QueryTimeout` if no reply arrives within ``timeout`` seconds;
        per §7.3 the handler decides whether to abort the stream or proceed.
        """
        if isinstance(prompt, str):
            prompt_text = prompt
            query_attachments = list(attachments) if attachments else None
        elif isinstance(prompt, Envelope):
            prompt_text = prompt.prompt
            merged = list(prompt.attachments or [])
            if attachments:
                merged.extend(attachments)
            query_attachments = merged or None
        else:
            raise TypeError(f"unsupported prompt type: {type(prompt).__name__}")

        reply_subject = self._nc.new_inbox()
        sub = await self._nc.subscribe(reply_subject)
        query = QueryChunk(
            id=uuid.uuid4().hex,
            reply_subject=reply_subject,
            prompt=prompt_text,
            attachments=query_attachments,
        )
        try:
            await self.send(query)
            try:
                msg = await sub.next_msg(timeout=timeout)
            except TimeoutError as exc:
                raise QueryTimeout(
                    f"no reply on {reply_subject} within {timeout}s (query id={query.id})"
                ) from exc
            return decode(msg.data)
        finally:
            with contextlib.suppress(Exception):
                await sub.unsubscribe()


class AgentService:
    """A protocol-compliant agent (§12 implementation checklist).

    Construct with ``agent``/``owner``/``name`` plus a live NATS client,
    register a prompt handler via :meth:`on_prompt`, then call
    :meth:`start`. The agent keeps a background heartbeat task running
    until :meth:`stop` is awaited.

    ``heartbeat_interval_s`` defaults to 30 s per §8.2. ``session`` is
    carried in both service metadata (§3.2) and the heartbeat payload
    (§8.3) when set; session-aware harnesses (``claude-code``, ``pi``,
    ``hermes``) MUST supply it.

    ``keepalive_interval_s`` controls per-request keep-alive: while a
    handler is running, the agent emits ``{"type":"status","data":"ack"}``
    every ``keepalive_interval_s`` seconds so callers using a stream
    inactivity timeout (the TS SDK default is 60 s) don't fire on slow
    handlers. Defaults to 30 s, matching the TS reference harnesses.
    Pass ``None`` to disable — for example when the handler emits its
    own status chunks at a finer cadence.
    """

    def __init__(
        self,
        *,
        agent: str,
        owner: str,
        name: str,
        nc: NATSClient,
        description: str = "",
        heartbeat_interval_s: int = 30,
        max_payload: str = DEFAULT_MAX_PAYLOAD,
        attachments_ok: bool = DEFAULT_ATTACHMENTS_OK,
        session: str | None = None,
        keepalive_interval_s: float | None = DEFAULT_KEEPALIVE_INTERVAL_S,
    ) -> None:
        if heartbeat_interval_s <= 0:
            raise ValueError("heartbeat_interval_s must be > 0 (heartbeat is mandatory in v0.3)")
        if keepalive_interval_s is not None and keepalive_interval_s <= 0:
            raise ValueError("keepalive_interval_s must be > 0 or None (None disables keep-alive)")
        # Validate max_payload eagerly so misconfiguration fails at construction
        # rather than surfacing later via caller-side validation (§5.4).
        parse_human_bytes(max_payload)
        self.subject = AgentSubject.new(agent=agent, owner=owner, name=name)
        self._nc = nc
        self._description = description or f"{agent} agent {self.subject.name}"
        self._session = session
        self._heartbeat_interval_s = heartbeat_interval_s
        self._max_payload = max_payload
        self._attachments_ok = attachments_ok
        self._keepalive_interval_s = keepalive_interval_s
        self._prompt_handler: PromptHandler | None = None
        self._service: Service | None = None
        self._heartbeat_task: asyncio.Task[None] | None = None
        self._heartbeat_stop = asyncio.Event()

    def on_prompt(self, handler: PromptHandler) -> None:
        """Register the prompt handler. Must be called before :meth:`start`."""
        self._prompt_handler = handler

    async def start(self) -> None:
        if self._prompt_handler is None:
            raise RuntimeError("register a prompt handler via on_prompt() before start()")
        if self._service is not None:
            raise RuntimeError("agent already started")

        metadata: dict[str, str] = {
            "agent": self.subject.agent,
            "owner": self.subject.owner,
            "protocol_version": _PROTOCOL_VERSION,
        }
        if self._session is not None:
            metadata["session"] = self._session
        config = ServiceConfig(
            name=SERVICE_NAME,
            version=_SDK_VERSION,
            description=self._description,
            metadata=metadata,
        )
        self._service = await add_service(self._nc, config)
        await self._service.add_endpoint(
            EndpointConfig(
                name=PROMPT_ENDPOINT_NAME,
                subject=self.subject.prompt,
                handler=self._on_prompt_request,
                # §3.3: the `prompt` endpoint MUST use queue group `"agents"`
                # so multiple instances of the same logical agent load-balance
                # requests. Framework defaults differ between SDKs, which
                # would break interop — pin to the spec value explicitly.
                queue_group=PROMPT_QUEUE_GROUP,
                # §2.1: endpoint metadata is a `Record<string, string>` on the
                # wire — booleans are encoded as "true" / "false".
                metadata={
                    "max_payload": self._max_payload,
                    "attachments_ok": "true" if self._attachments_ok else "false",
                },
            )
        )
        # v0.3 §-TBD: the status endpoint returns a freshly-built heartbeat-
        # shaped payload. Same queue group as `prompt` so callers load-balance
        # to one responder per logical agent.
        await self._service.add_endpoint(
            EndpointConfig(
                name=STATUS_ENDPOINT_NAME,
                subject=self.subject.status,
                handler=self._on_status_request,
                queue_group=STATUS_QUEUE_GROUP,
            )
        )

        self._heartbeat_stop.clear()
        # §8.3 instance_id matches the micro-service instance id assigned by
        # nats-py; this is what callers correlate heartbeats against.
        self._heartbeat_task = asyncio.create_task(
            run_publisher(
                self._nc,
                self.subject,
                self._heartbeat_interval_s,
                self._service.id,
                self._heartbeat_stop,
                session=self._session,
            ),
            name=f"heartbeat-{self.subject.inbox}",
        )
        log.info("agent started on %s (instance_id=%s)", self.subject.inbox, self._service.id)

    async def stop(self) -> None:
        self._heartbeat_stop.set()
        if self._heartbeat_task is not None:
            with contextlib.suppress(asyncio.CancelledError):
                await self._heartbeat_task
            self._heartbeat_task = None
        if self._service is not None:
            await self._service.stop()
            self._service = None
        log.info("agent stopped on %s", self.subject.inbox)

    async def _on_status_request(self, request: Request) -> None:
        """Reply with a freshly-built §8.3 heartbeat payload (v0.3 §-TBD).

        Request body is ignored — there is no request schema yet; future
        PRs will extend the response with richer metadata and may add a
        request shape at that time. ``self._service`` is set in
        :meth:`start` before the endpoint is registered, so the
        ``RuntimeError`` guard below should be unreachable in normal
        operation; it survives ``python -O`` (which strips ``assert``)
        and gives a clearer signal than ``AttributeError`` from the
        following ``.id`` access if invariants ever drift.
        """
        if self._service is None:  # pragma: no cover — defensive
            raise RuntimeError("status handler invoked before start()")
        try:
            payload = build_heartbeat_payload(
                self.subject,
                self._heartbeat_interval_s,
                self._service.id,
                self._session,
            )
            data = payload.model_dump_json(exclude_none=True).encode("utf-8")
            await request.respond(data)
        except Exception as exc:
            # A respond() failure (broker dropped, request torn down, encode
            # error in a future richer payload) MUST surface as a §9.1 error
            # to the caller, not silently propagate into nats-py's framework
            # — mirroring _on_prompt_request's explicit error path.
            log.exception("status handler failed on %s", request.subject)
            with contextlib.suppress(Exception):
                await request.respond_error(
                    "500", _sanitize_error_desc(f"status handler error: {exc}")
                )

    async def _on_prompt_request(self, request: Request) -> None:
        keepalive_task: asyncio.Task[None] | None = None
        try:
            try:
                envelope = decode(request.data)
            except ProtocolError as exc:
                log.warning("rejecting malformed prompt on %s: %s", request.subject, exc)
                await request.respond_error("400", _sanitize_error_desc(str(exc)))
                return

            stream = PromptStream(request, self._nc)
            handler = self._prompt_handler
            if handler is None:  # pragma: no cover — start() rejects this path
                raise RuntimeError("prompt handler invoked before on_prompt() registered one")

            if self._keepalive_interval_s is not None:
                keepalive_task = asyncio.create_task(
                    _keepalive_loop(request, self._keepalive_interval_s),
                    name=f"keepalive-{request.subject}",
                )

            try:
                await handler(envelope, stream)
            except Exception as exc:
                log.exception("prompt handler raised on %s", request.subject)
                # Stop keep-alive BEFORE the §9 error frame so the keepalive
                # task can't race an ack chunk in between error(500) and the
                # §6.5 terminator emitted in the outer `finally`. Ditto in
                # the success path right below.
                await _stop_keepalive(keepalive_task)
                keepalive_task = None
                await request.respond_error("500", _sanitize_error_desc(f"handler error: {exc}"))
            else:
                await _stop_keepalive(keepalive_task)
                keepalive_task = None
        finally:
            # Belt-and-braces: if anything above failed before we could stop
            # the keepalive (e.g. respond_error itself raised), don't leak the
            # task and don't let it slip an ack past the terminator. No-op
            # when keepalive_task is already None.
            await _stop_keepalive(keepalive_task)
            # §6.5 + §9.3: every stream — successful or errored — ends with a
            # zero-byte body message that carries NO NATS headers. The error
            # frame emitted by `respond_error` above is NOT the terminator;
            # this final `respond(b"")` is.
            try:
                await request.respond(b"")
            except Exception:
                log.exception("failed to emit stream terminator on %s", request.subject)


async def _stop_keepalive(task: asyncio.Task[None] | None) -> None:
    """Cancel and await a keep-alive task, swallowing the resulting CancelledError.

    No-op when ``task`` is ``None``, so the outer ``finally`` can call this
    unconditionally as a belt-and-braces guard after the success/error paths
    have already cleared their own task handles.
    """
    if task is None:
        return
    task.cancel()
    with contextlib.suppress(asyncio.CancelledError):
        await task


async def _keepalive_loop(request: Request, interval_s: float) -> None:
    """Emit a `status="ack"` chunk every `interval_s` seconds until cancelled.

    Runs as a sibling task to the user's prompt handler; cancelled by
    :meth:`AgentService._on_prompt_request` once the handler returns or
    raises, so the ack stream never extends past the §6.5 terminator.
    """
    payload = encode_chunk(StatusChunk(status="ack"))
    while True:
        await asyncio.sleep(interval_s)
        try:
            await request.respond(payload)
        except Exception:
            # An emit failure (broker dropped, request reply already torn down,
            # etc.) is best-effort — log and stop. The terminator path will
            # fail loudly enough on its own if the request is truly dead.
            log.exception("keepalive emit failed on %s", request.subject)
            return


# NATS message headers are single-line (CR/LF delimited in the wire format),
# so any description passed to `respond_error` MUST be stripped of newlines
# or the server will truncate subsequent headers. 200 chars is plenty for
# §9.1 ("short human-readable description"); richer context belongs in the
# JSON body per §9.1.
_MAX_ERROR_DESC_LEN = 200


def _sanitize_error_desc(desc: str) -> str:
    flat = " | ".join(line.strip() for line in desc.splitlines() if line.strip())
    if len(flat) > _MAX_ERROR_DESC_LEN:
        # ASCII "..." (not U+2026) — some NATS header parsers choke on multi-
        # byte UTF-8 in header values.
        flat = flat[: _MAX_ERROR_DESC_LEN - 3] + "..."
    return flat


__all__ = [
    "DEFAULT_ATTACHMENTS_OK",
    "DEFAULT_KEEPALIVE_INTERVAL_S",
    "DEFAULT_MAX_PAYLOAD",
    "AgentService",
    "PromptHandler",
    "PromptStream",
]
